bbcondeparser/html_tags.py

```diff
@@ -358,108 +343,235 @@ class BaseHTMLRenderTreeParser(BaseTreeParser):
-                    rendered_children.append(node.render_raw())
+    def __init__(self):
+        self.stack = []
 
+    def __bool__(self):
+        return len(self) > 0
+
+    __nonzero__ = __bool__
+
+    def __len__(self):
+        return len(self.stack)
+
+    def __getitem__(self, index):
+        return self.stack[index]
+
+    def pop(self):
+        return self.stack.pop()
+
+    def push(self, *args, **kwargs):
+        item = StackLevel(*args, **kwargs)
+        self.stack.append(item)
+
+
+class _BaseHTMLRenderTreeParser(object):
+
+    def __init__(self, root_node, newline_behaviour=None,
+                 convert_paragraphs=None):
+
+        self.root_node = root_node
+        self.stack = NodeStack()
+        self.newline_behaviour = newline_behaviour
+        self.convert_paragraphs = convert_paragraphs
+
+        self._inside_paragraph = False
+        self._paragraph_tree = None
+        self._tree = []
+        self._node = None
+
+    def append_tree(self, item):
+        self._tree.append(item)
+
+    def append_paragraph(self, item):
+        self._paragraph_tree.append(item)
+
+    def stack_push(self):
+        self.stack.push(
+            self._tree, self._node, self._paragraph_tree,
+            self._inside_paragraph, self.newline_behaviour,
+            self.convert_paragraphs,
+        )
+        self._tree = []
+        # self._node = None
+        self._paragraph_tree = None
+        self._inside_paragraph = False
+        ctx = self._node.get_context()
+        self.newline_behaviour = get_newline_behaviour(
+            self.newline_behaviour, ctx.get('newline_behaviour'))
+        self.convert_paragraphs = get_convert_paragraphs(
+            self.convert_paragraphs, ctx.get('convert_paragraphs'))
+
+    def stack_pop(self):
+        self.set_state(self.stack.pop())
+
+    def set_state(self, stack_ctx):
+        self._tree = stack_ctx.tree
+        self._node = stack_ctx.node
+        self._paragraph_tree = stack_ctx.paragraph_tree
+        self._inside_paragraph = stack_ctx.inside_paragraph
+        self.newline_behaviour = stack_ctx.newline_behaviour
+        self.convert_paragraphs = stack_ctx.convert_paragraphs
+
+    @property
+    def is_converting_paragraphs(self):
+        return self.convert_paragraphs is True
+
+    @property
+    def is_removing_newlines(self):
+        return self.newline_behaviour == 'remove'
+
+    @property
+    def is_converting_newlines(self):
+        return self.newline_behaviour == 'convert'
+
+    @property
+    def is_inside_paragraph(self):
+        return self._inside_paragraph is True
+
+    def open_paragraph(self, include_node=True):
+        self._inside_paragraph = True
+        if include_node:
+            self._paragraph_tree = [self._node]
+        else:
+            self._paragraph_tree = []
+
+    def close_paragraph(self):
+        # clone paragraph scope?
+        paragraph_node = ParagraphTag({}, self._paragraph_tree, '', '')
```

Comment:
Could we add a comment here to remind the reader that the `ParagraphTag` is going to set itself as the `parent` on the children in `self._paragraph_tree`?